atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/TargetedSweeper.java

```diff
@@ -149,17 +158,39 @@ private void scheduleBackgroundThreads() {
                 executorService = PTExecutors
                         .newScheduledThreadPoolExecutor(numThreads, new NamedThreadFactory("Targeted Sweep", true));
                 for (int i = 0; i < numThreads; i++) {
-                    executorService.scheduleWithFixedDelay(
-                            () -> sweepNextBatch(ShardAndStrategy.of(getShardAndIncrement(), sweepStrategy)),
-                            1, 5, TimeUnit.SECONDS);
+                    executorService.scheduleWithFixedDelay(this::runOneIteration, 1, 5, TimeUnit.SECONDS);
                 }
             }
         }
 
+        private void runOneIteration() {
+            ShardAndStrategy shardStrategy = ShardAndStrategy.of(lockNextShardToSweep(), sweepStrategy);
+            try {
+                sweepNextBatch(shardStrategy);
+            } catch (Throwable th) {
+                log.warn("Targeted sweep for {} failed and will be retried later.",
+                        SafeArg.of("shardStrategy", shardStrategy.toText()), th);
+            } finally {
+                unlockShard(shardStrategy.shard());
+            }
+        }
+
+        private int lockNextShardToSweep() {
+            int nextShardCandidate = getShardAndIncrement();
+            while (!shardsBeingSwept.add(nextShardCandidate)) {
+                nextShardCandidate = getShardAndIncrement();
+            }
+            return nextShardCandidate;
```

Comment:
Discussed offline - might be a little awkward if we have more threads than shards, but we could at initialisation transform the numbers so things work